File: CS/SaveWpfGridState/RowStateHelper.cs
// Developer Express Code Central Example:
// How to preserve the DXGrid for WPF view state
// 
// This example demonstrates how to save/load the groups and master rows expansion
// state, selection, focused row and position of the focused row.
// All
// functionality implemented in custom class named RowStateHelper, that contains
// methods for performing these operations.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4222

using DevExpress.Xpf.Grid;
using System.Collections;
using System.Collections.Generic;

namespace SaveWpfGridState {
    public class RowStateHelper {
        GridControl gridControl;
        string keyFieldName;

        object focusedMasterKey;
        object focusedDetailKey;
        List<object> focusedGroupValues;

        ArrayList selectionList;
        ArrayList masterRowList;
        List<GroupInfo> groupRowList;

        public ArrayList SelectionList
        {
            get
            {
                if (selectionList == null)
                    selectionList = new ArrayList();
                return selectionList;
            }
        }
        public ArrayList MasterRowList
        {
            get
            {
                if (masterRowList == null)
                    masterRowList = new ArrayList();
                return masterRowList;
            }
        }
        public List<GroupInfo> GroupRowList
        {
            get
            {
                if (groupRowList == null)
                    groupRowList = new List<GroupInfo>();
                return groupRowList;
            }
        }

        public RowStateHelper(GridControl vw, string keyField) {
            gridControl = vw;
            keyFieldName = keyField;
        }

        public void SaveViewInfo() {
            SaveExpandedMasterRows();
            SaveExpansionView();
            //SaveSelectionView();
            SaveSelection();
        }
        public void LoadViewInfo() {
            LoadExpandedMasterRows();
            LoadExpansionView();
            //LoadSelectionView();
            LoadSelection();
            ClearVariables();
        }
        void ClearVariables() {
            focusedMasterKey = null;
            focusedDetailKey = null;
            focusedGroupValues = null;
            SelectionList.Clear();
            MasterRowList.Clear();
            GroupRowList.Clear();
        }

        #region master rows
        void SaveExpandedMasterRows() {
            MasterRowList.Clear();
            for (int i = 0; i < gridControl.DataController.ListSourceRowCount; i++) {
                var rowHandle = gridControl.GetRowHandleByListIndex(i);
                if (gridControl.IsMasterRowExpanded(rowHandle))
                    MasterRowList.Add(GetKeyByRowHandle(rowHandle));
            }
        }
        void LoadExpandedMasterRows() {
            for (int i = 0; i < gridControl.DataController.ListSourceRowCount; i++) {
                var rowHandle = gridControl.GetRowHandleByListIndex(i);
                gridControl.CollapseMasterRow(rowHandle);
            }
            foreach (var key in MasterRowList) {
                var test = GetRowHandleByKey(key);
                var datarow = gridControl.DataController.GetRow(0);
                gridControl.ExpandMasterRow(GetRowHandleByKey(key));
            }
        }
        #endregion

        #region group rows
        void SaveExpansionView() {
            if (gridControl.GroupCount == 0)
                return;
            GroupRowList.Clear();
            for (int i = -1; i > int.MinValue; i--) {
                if (!gridControl.IsValidRowHandle(i))
                    break;
                if (gridControl.GetRowLevelByRowHandle(i) == 0)
                    GroupRowList.Add(GetGroupInfo(i));
            }
        }
        GroupInfo GetGroupInfo(int rowHandle) {
            var info = new GroupInfo {
                Value = gridControl.GetGroupRowValue(rowHandle),
                IsExpanded = gridControl.IsGroupRowExpanded(rowHandle)
            };
            for (var i = 0; i < gridControl.GetChildRowCount(rowHandle); i++) {
                var childRowHandle = gridControl.GetChildRowHandle(rowHandle, i);
                if (gridControl.IsGroupRowHandle(childRowHandle))
                    info.Children.Add(GetGroupInfo(childRowHandle));
            }
            return info;
        }

        void LoadExpansionView() {
            if (gridControl.GroupCount == 0)
                return;
            gridControl.CollapseAllGroups();
            for (int i = -1; i > int.MinValue; i--) {
                if (!gridControl.IsValidRowHandle(i))
                    break;
                if (gridControl.GetRowLevelByRowHandle(i) == 0)
                    ExpandGroupRow(i, GroupRowList);
            }
        }
        void ExpandGroupRow(int rowHandle, List<GroupInfo> infoList) {
            var value = gridControl.GetGroupRowValue(rowHandle);
            var info = infoList.Find(i => i.Value.Equals(value));
            if (info == null)
                return;
            if (info.IsExpanded)
                gridControl.ExpandGroupRow(rowHandle);
            for (int i = 0; i < gridControl.GetChildRowCount(rowHandle); i++) {
                var childRowHandle = gridControl.GetChildRowHandle(rowHandle, i);
                if (gridControl.IsGroupRowHandle(childRowHandle))
                    ExpandGroupRow(childRowHandle, info.Children);
            }
            infoList.Remove(info);
        }
        #endregion

        #region focused row
        void SaveSelection() {
            focusedDetailKey = null;
            focusedMasterKey = null;
            focusedGroupValues = null;

            int focusedRowHandle;
            if (gridControl.View.IsFocusedView) {
                focusedDetailKey = null;
                focusedRowHandle = gridControl.View.FocusedRowHandle;
                if (gridControl.IsGroupRowHandle(focusedRowHandle)) {
                    focusedGroupValues = GetGroupValues(focusedRowHandle);
                    return;
                }
            }
            else {
                var focusedView = gridControl.View.FocusedView;
                var detailRowHandle = focusedView.FocusedRowHandle;
                focusedDetailKey = GetKeyByRowHandle(detailRowHandle, (GridControl)focusedView.DataControl);
                focusedRowHandle = ((GridControl)focusedView.DataControl).GetMasterRowHandle();
            }
            focusedMasterKey = GetKeyByRowHandle(focusedRowHandle);
        }

        void LoadSelection() {
            var masterRowHandle = (focusedGroupValues == null) ? GetRowHandleByKey(focusedMasterKey) : GetGroupRowHandleByValue();
            if (focusedDetailKey == null) {
                gridControl.View.FocusedRowHandle = masterRowHandle;
                return;
            }
            if (!gridControl.IsMasterRowExpanded(masterRowHandle))
                gridControl.ExpandMasterRow(masterRowHandle);
            var detailControl = gridControl.GetDetail(masterRowHandle) as GridControl;
            if (detailControl == null)
                return;
            detailControl.View.MoveFocusedRow(GetRowHandleByKey(focusedDetailKey, detailControl));
        }

        List<object> GetGroupValues(int rowHandle) {
            var list = new List<object>();
            while (rowHandle != DataControlBase.InvalidRowHandle) {
                list.Insert(0, gridControl.GetGroupRowValue(rowHandle));
                rowHandle = gridControl.GetParentRowHandle(rowHandle);
            }
            return list;
        }

        int GetGroupRowHandleByValue() {
            if (focusedGroupValues.Count == 0)
                return DataControlBase.InvalidRowHandle;
            for (int i = -1; gridControl.IsValidRowHandle(i); i--) {
                if (gridControl.GetRowLevelByRowHandle(i) == 0) {
                    var rowHandle = FindGroupRowHandle(i, 0);
                    if (rowHandle != DataControlBase.InvalidRowHandle)
                        return rowHandle;
                }
            }
            return DataControlBase.InvalidRowHandle;
        }
        int FindGroupRowHandle(int rowHandle, int level) {
            var value = gridControl.GetGroupRowValue(rowHandle);
            if (!value.Equals(focusedGroupValues[level]))
                return DataControlBase.InvalidRowHandle;
            if (focusedGroupValues.Count - 1 != level) {
                for (int i = 0; i < gridControl.GetChildRowCount(rowHandle); i++) {
                    var childRowHandle = gridControl.GetChildRowHandle(rowHandle, i);
                    if (!gridControl.IsGroupRowHandle(childRowHandle))
                        continue;
                    var result = FindGroupRowHandle(childRowHandle, level + 1);
                    if (result != DataControlBase.InvalidRowHandle)
                        return result;
                }
            }
            return rowHandle;
        }
        #endregion

        #region selection
        void SaveSelectionView() {
            SelectionList.Clear();
            var selectedRows = gridControl.GetSelectedRowHandles();
            for (int i = 0; i < selectedRows.Length; i++) {
                SelectionList.Add(GetKeyByRowHandle(selectedRows[i]));
            }
        }

        void LoadSelectionView() {
            gridControl.BeginSelection();
            try {
                gridControl.UnselectAll();
                for (int i = 0; i < SelectionList.Count; i++)
                    gridControl.SelectItem(GetRowHandleByKey(SelectionList[i]));
            }
            finally {
                gridControl.EndSelection();
            }
        }
        #endregion

        object GetKeyByRowHandle(int rowHandle) {
            return GetKeyByRowHandle(rowHandle, gridControl);
        }
        object GetKeyByRowHandle(int rowHandle, GridControl grid) {
            return grid.GetCellValue(rowHandle, keyFieldName);
        }

        int GetRowHandleByKey(object key) {
            return GetRowHandleByKey(key, gridControl);
        }
        int GetRowHandleByKey(object key, GridControl grid) {
            return grid.FindRowByValue(keyFieldName, key);
        }
    }
    public class GroupInfo {
        public GroupInfo() {
            Children = new List<GroupInfo>();
        }
        public object Value { get; set; }
        public bool IsExpanded { get; set; }
        public List<GroupInfo> Children { get; private set; }
    }
}

File: VB/SaveWpfGridState/RowStateHelper.vb
' Developer Express Code Central Example:
' How to preserve the DXGrid for WPF view state
' 
' This example demonstrates how to save/load the groups and master rows expansion
' state, selection, focused row and position of the focused row.
' All
' functionality implemented in custom class named RowStateHelper, that contains
' methods for performing these operations.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E4222

Imports DevExpress.Xpf.Grid
Imports System.Collections
Imports System.Collections.Generic

Namespace SaveWpfGridState
    Public Class RowStateHelper
        Private gridControl As GridControl
        Private keyFieldName As String

        Private focusedMasterKey As Object
        Private focusedDetailKey As Object
        Private focusedGroupValues As List(Of Object)


        Private selectionList_Renamed As ArrayList

        Private masterRowList_Renamed As ArrayList

        Private groupRowList_Renamed As List(Of GroupInfo)

        Public ReadOnly Property SelectionList() As ArrayList
            Get
                If selectionList_Renamed Is Nothing Then
                    selectionList_Renamed = New ArrayList()
                End If
                Return selectionList_Renamed
            End Get
        End Property
        Public ReadOnly Property MasterRowList() As ArrayList
            Get
                If masterRowList_Renamed Is Nothing Then
                    masterRowList_Renamed = New ArrayList()
                End If
                Return masterRowList_Renamed
            End Get
        End Property
        Public ReadOnly Property GroupRowList() As List(Of GroupInfo)
            Get
                If groupRowList_Renamed Is Nothing Then
                    groupRowList_Renamed = New List(Of GroupInfo)()
                End If
                Return groupRowList_Renamed
            End Get
        End Property

        Public Sub New(ByVal vw As GridControl, ByVal keyField As String)
            gridControl = vw
            keyFieldName = keyField
        End Sub

        Public Sub SaveViewInfo()
            SaveExpandedMasterRows()
            SaveExpansionView()
            'SaveSelectionView();
            SaveSelection()
        End Sub
        Public Sub LoadViewInfo()
            LoadExpandedMasterRows()
            LoadExpansionView()
            'LoadSelectionView();
            LoadSelection()
            ClearVariables()
        End Sub
        Private Sub ClearVariables()
            focusedMasterKey = Nothing
            focusedDetailKey = Nothing
            focusedGroupValues = Nothing
            SelectionList.Clear()
            MasterRowList.Clear()
            GroupRowList.Clear()
        End Sub

        #Region "master rows"
        Private Sub SaveExpandedMasterRows()
            MasterRowList.Clear()
            For i As Integer = 0 To gridControl.DataController.ListSourceRowCount - 1
                Dim rowHandle = gridControl.GetRowHandleByListIndex(i)
                If gridControl.IsMasterRowExpanded(rowHandle) Then
                    MasterRowList.Add(GetKeyByRowHandle(rowHandle))
                End If
            Next i
        End Sub
        Private Sub LoadExpandedMasterRows()
            For i As Integer = 0 To gridControl.DataController.ListSourceRowCount - 1
                Dim rowHandle = gridControl.GetRowHandleByListIndex(i)
                gridControl.CollapseMasterRow(rowHandle)
            Next i
            For Each key In MasterRowList
                Dim test = GetRowHandleByKey(key)
                Dim datarow = gridControl.DataController.GetRow(0)
                gridControl.ExpandMasterRow(GetRowHandleByKey(key))
            Next key
        End Sub
        #End Region

        #Region "group rows"
        Private Sub SaveExpansionView()
            If gridControl.GroupCount = 0 Then
                Return
            End If
            GroupRowList.Clear()
            For i As Integer = -1 To Integer.MinValue + 1 Step -1
                If Not gridControl.IsValidRowHandle(i) Then
                    Exit For
                End If
                If gridControl.GetRowLevelByRowHandle(i) = 0 Then
                    GroupRowList.Add(GetGroupInfo(i))
                End If
            Next i
        End Sub
        Private Function GetGroupInfo(ByVal rowHandle As Integer) As GroupInfo
            Dim info = New GroupInfo With { _
                .Value = gridControl.GetGroupRowValue(rowHandle), _
                .IsExpanded = gridControl.IsGroupRowExpanded(rowHandle) _
            }
            Dim i = 0
            Do While i < gridControl.GetChildRowCount(rowHandle)
                Dim childRowHandle = gridControl.GetChildRowHandle(rowHandle, i)
                If gridControl.IsGroupRowHandle(childRowHandle) Then
                    info.Children.Add(GetGroupInfo(childRowHandle))
                End If
                i += 1
            Loop
            Return info
        End Function

        Private Sub LoadExpansionView()
            If gridControl.GroupCount = 0 Then
                Return
            End If
            gridControl.CollapseAllGroups()
            For i As Integer = -1 To Integer.MinValue + 1 Step -1
                If Not gridControl.IsValidRowHandle(i) Then
                    Exit For
                End If
                If gridControl.GetRowLevelByRowHandle(i) = 0 Then
                    ExpandGroupRow(i, GroupRowList)
                End If
            Next i
        End Sub
        Private Sub ExpandGroupRow(ByVal rowHandle As Integer, ByVal infoList As List(Of GroupInfo))
            Dim value = gridControl.GetGroupRowValue(rowHandle)
            Dim info = infoList.Find(Function(x) Equals(x.Value, value))
            If info Is Nothing Then
                Return
            End If
            If info.IsExpanded Then
                gridControl.ExpandGroupRow(rowHandle)
            End If
            Dim i As Integer = 0
            Do While i < gridControl.GetChildRowCount(rowHandle)
                Dim childRowHandle = gridControl.GetChildRowHandle(rowHandle, i)
                If gridControl.IsGroupRowHandle(childRowHandle) Then
                    ExpandGroupRow(childRowHandle, info.Children)
                End If
                i += 1
            Loop
            infoList.Remove(info)
        End Sub
        #End Region

        #Region "focused row"
        Private Sub SaveSelection()
            focusedDetailKey = Nothing
            focusedMasterKey = Nothing
            focusedGroupValues = Nothing

            Dim focusedRowHandle As Integer
            If gridControl.View.IsFocusedView Then
                focusedDetailKey = Nothing
                focusedRowHandle = gridControl.View.FocusedRowHandle
                If gridControl.IsGroupRowHandle(focusedRowHandle) Then
                    focusedGroupValues = GetGroupValues(focusedRowHandle)
                    Return
                End If
            Else
                Dim focusedView = gridControl.View.FocusedView
                Dim detailRowHandle = focusedView.FocusedRowHandle
                focusedDetailKey = GetKeyByRowHandle(detailRowHandle, CType(focusedView.DataControl, GridControl))
                focusedRowHandle = CType(focusedView.DataControl, GridControl).GetMasterRowHandle()
            End If
            focusedMasterKey = GetKeyByRowHandle(focusedRowHandle)
        End Sub

        Private Sub LoadSelection()
            Dim masterRowHandle = If(focusedGroupValues Is Nothing, GetRowHandleByKey(focusedMasterKey), GetGroupRowHandleByValue())
            If focusedDetailKey Is Nothing Then
                gridControl.View.FocusedRowHandle = masterRowHandle
                Return
            End If
            If Not gridControl.IsMasterRowExpanded(masterRowHandle) Then
                gridControl.ExpandMasterRow(masterRowHandle)
            End If
            Dim detailControl = TryCast(gridControl.GetDetail(masterRowHandle), GridControl)
            If detailControl Is Nothing Then
                Return
            End If
            detailControl.View.MoveFocusedRow(GetRowHandleByKey(focusedDetailKey, detailControl))
        End Sub

        Private Function GetGroupValues(ByVal rowHandle As Integer) As List(Of Object)
            Dim list = New List(Of Object)()
            Do While rowHandle <> DataControlBase.InvalidRowHandle
                list.Insert(0, gridControl.GetGroupRowValue(rowHandle))
                rowHandle = gridControl.GetParentRowHandle(rowHandle)
            Loop
            Return list
        End Function

        Private Function GetGroupRowHandleByValue() As Integer
            If focusedGroupValues.Count = 0 Then
                Return DataControlBase.InvalidRowHandle
            End If
            Dim i As Integer = -1
            Do While gridControl.IsValidRowHandle(i)
                If gridControl.GetRowLevelByRowHandle(i) = 0 Then
                    Dim rowHandle = FindGroupRowHandle(i, 0)
                    If rowHandle <> DataControlBase.InvalidRowHandle Then
                        Return rowHandle
                    End If
                End If
                i -= 1
            Loop
            Return DataControlBase.InvalidRowHandle
        End Function
        Private Function FindGroupRowHandle(ByVal rowHandle As Integer, ByVal level As Integer) As Integer
            Dim value = gridControl.GetGroupRowValue(rowHandle)
            If Not Equals(value, focusedGroupValues(level)) Then
                Return DataControlBase.InvalidRowHandle
            End If
            If focusedGroupValues.Count - 1 <> level Then
                Dim i As Integer = 0
                Do While i < gridControl.GetChildRowCount(rowHandle)
                    Dim childRowHandle = gridControl.GetChildRowHandle(rowHandle, i)
                    If Not gridControl.IsGroupRowHandle(childRowHandle) Then
                        i += 1
                        Continue Do
                    End If
                    Dim result = FindGroupRowHandle(childRowHandle, level + 1)
                    If result <> DataControlBase.InvalidRowHandle Then
                        Return result
                    End If
                    i += 1
                Loop
            End If
            Return rowHandle
        End Function
        #End Region

        #Region "selection"
        Private Sub SaveSelectionView()
            SelectionList.Clear()
            Dim selectedRows = gridControl.GetSelectedRowHandles()
            For i As Integer = 0 To selectedRows.Length - 1
                SelectionList.Add(GetKeyByRowHandle(selectedRows(i)))
            Next i
        End Sub

        Private Sub LoadSelectionView()
            gridControl.BeginSelection()
            Try
                gridControl.UnselectAll()
                For i As Integer = 0 To SelectionList.Count - 1
                    gridControl.SelectItem(GetRowHandleByKey(SelectionList(i)))
                Next i
            Finally
                gridControl.EndSelection()
            End Try
        End Sub
        #End Region

        Private Function GetKeyByRowHandle(ByVal rowHandle As Integer) As Object
            Return GetKeyByRowHandle(rowHandle, gridControl)
        End Function
        Private Function GetKeyByRowHandle(ByVal rowHandle As Integer, ByVal grid As GridControl) As Object
            Return grid.GetCellValue(rowHandle, keyFieldName)
        End Function

        Private Function GetRowHandleByKey(ByVal key As Object) As Integer
            Return GetRowHandleByKey(key, gridControl)
        End Function
        Private Function GetRowHandleByKey(ByVal key As Object, ByVal grid As GridControl) As Integer
            Return grid.FindRowByValue(keyFieldName, key)
        End Function
    End Class
    Public Class GroupInfo
        Public Property Value() As Object
        Public Property IsExpanded() As Boolean
        Private privateChildren As New List(Of GroupInfo)()
        Public Property Children() As List(Of GroupInfo)
            Get
                Return privateChildren
            End Get
            Private Set(ByVal value As List(Of GroupInfo))
                privateChildren = value
            End Set
        End Property
    End Class
End Namespace
